pkg/k2/constructs/dynamic_value.go

```diff
@@ -0,0 +1,325 @@
+	if err != nil {
+		return "", err
+	}
+	return string(j), nil
+}
+
+func (ctx DynamicValueContext) PathAncestor(path construct.PropertyPath, depth int) (string, error) {
+	if depth < 0 {
+		return "", fmt.Errorf("depth must be >= 0")
+	}
+	if depth == 0 {
+		return path.String(), nil
+	}
+	if len(path) <= depth {
+		return "", fmt.Errorf("depth %d is greater than path length %d", depth, len(path))
+	}
+	return path[:len(path)-depth].String(), nil
+}
+
+func (ctx DynamicValueContext) PathAncestorExists(path construct.PropertyPath, depth int) bool {
+	return len(path) > depth
+}
+
+// FieldRef returns a reference to `field` on `resource` (as a PropertyRef)
+func (ctx DynamicValueContext) FieldRef(field string, resource any) (construct.PropertyRef, error) {
+	resId, err := TemplateArgToRID(resource)
+	if err != nil {
+		return construct.PropertyRef{}, err
+	}
+
+	return construct.PropertyRef{
+		Resource: resId,
+		Property: field,
+	}, nil
+}
+
+func TemplateArgToRID(arg any) (construct.ResourceId, error) {
+	switch arg := arg.(type) {
+	case construct.ResourceId:
+		return arg, nil
+
+	case construct.Resource:
+		return arg.ID, nil
+
+	case string:
+		var resId construct.ResourceId
+		err := resId.UnmarshalText([]byte(arg))
+		return resId, err
+	}
+
+	return construct.ResourceId{}, fmt.Errorf("invalid argument type %T", arg)
+}
+
+type DynamicValueSelection struct {
+	Source  any
+	mapKeys []reflect.Value
+	next    int
+	Key     string
+	Value   any
+	Index   int
+}
+
+func SelectItem(src any) DynamicValueSelection {
+	srcValue := reflect.ValueOf(src)
+	switch srcValue.Kind() {
+	case reflect.Map:
+		if !srcValue.IsValid() || srcValue.Len() == 0 {
+			return DynamicValueSelection{
+				Source: src,
+			}
+		}
+		keys := srcValue.MapKeys()
+		slices.SortStableFunc(keys, func(i, j reflect.Value) int {
+			return strings.Compare(stringValue(i.Interface()), stringValue(j.Interface()))
+		})
+		if len(keys) == 0 {
+			return DynamicValueSelection{
+				Source: src,
+			}
+		}
+		return DynamicValueSelection{
+			Source:  src,
+			mapKeys: keys,
+		}
+	default:
+		return DynamicValueSelection{
+			Source: src,
+		}
+	}
+}
+
+// Next returns the next value in the selection and whether there are more values
+func (s *DynamicValueSelection) Next() (any, bool) {
+	srcValue := reflect.ValueOf(s.Source)
+
+	if !srcValue.IsValid() {
+		return nil, false
+	}
+
+	if len(s.mapKeys) > 0 {
+		if s.next >= len(s.mapKeys) {
+			return nil, false
+		}
+		key := s.mapKeys[s.next]
+		value := srcValue.MapIndex(key).Interface()
+		s.Value = value
+		s.next++
+		return value, true
+	}
+	if s.Index >= srcValue.Len() {
+		return nil, false
+	}
+	value := srcValue.Index(s.Index).Interface()
+	s.Value = value
+	s.Index++
+	return value, true
+}
+
+func stringValue(v any) string {
+	if v == nil {
+		return ""
+	}
+	switch v := v.(type) {
+	case string:
+		return v
+	case fmt.Stringer:
+		return v.String()
+	default:
+		return fmt.Sprintf("%v", v)
+	}
```

Comment:
This switch is just a more complicated version of `fmt.Sprintf("%v", v)`. It's slightly faster, but I don't think it matters here.